### PR TITLE
[FEATURE] enhance model_uploader workflow to support MIT-licensed models from huggingface

### LINKED ISSUE #387: [FEATURE] enhance model_uploader workflow to support BGE models from huggingface
**Is your feature request related to a problem?**
In OpenSearch we support some sentence-transformers model as [pretrained models](https://opensearch.org/docs/latest/ml-commons-plugin/pretrained-models/#sentence-transformers). The registration of pretrained models is much more convenient, and users don't need to change the cluster settings plugins.ml_commons.allow_registering_model_via_url.

With the development of the research and engineering evolution in IR domain, now there are much stronger text_embedding models in the open source community. ([leaderboard ref](https://huggingface.co/spaces/mteb/leaderboard)) However, users still need to trace these models and generate the tarball manually, which is a heavy workload especially for those with little machine-learning background knowledge.

**What solution would you like?**
BGE models(https://huggingface.co/BAAI/bge-small-en-v1.5, https://huggingface.co/BAAI/bge-base-en-v1.5, https://huggingface.co/BAAI/bge-large-en-v1.5) have very strong text_embedding representation among the models with same size. And we can use them consistently with other sentence-transformers text_embedding models. 

Considering the models will consume resources in local deployment, We can support  bge-small-en-v1.5 and bge-base-en-v1.5 as pretrained models in OpenSearch.

**What alternatives have you considered?**

**Do you have any additional context?**
https://github.com/opensearch-project/ml-commons/issues/2210

### PR DESCRIPTION
**Pull request by zhichao-aws**
### Description
Code changes to enhance model upload workflow to support MIT licensed models from huggingface
 
### Issues Resolved
#387 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by zhichao-aws** [comment]
The workflow works well in my repo GH actions (I commented out the manual approve part)
They traced model and upload to S3 and create PR in my repo
link:
https://github.com/zhichao-aws/opensearch-py-ml/actions/runs/8843390731
https://github.com/zhichao-aws/opensearch-py-ml/actions/runs/8843323304

**Comment by zhichao-aws** [comment]
With this workflow, to upload BGE-base models, we trigger it with these settings, and the last input MIT license url should be : https://github.com/FlagOpen/FlagEmbedding/raw/master/LICENSE
image

**Comment by zhichao-aws** [comment]
> With this workflow, to upload BGE-small models, we trigger it with these settings, and the last input MIT license url should be : https://github.com/FlagOpen/FlagEmbedding/raw/master/LICENSE 

To upload BGE-small, change the model id to BAAI/bge-base-en-v1.5

**Comment by zhichao-aws** [comment]
One screenshot for the upload workflow at my GH repo
image

**Comment by zhichao-aws** [comment]
This PR is ready for review.

File: utils/model_uploader/third_party_statements.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    :rtype: str
+    """
+
+    r = requests.get(mit_license_url)
+    assert r.status_code == 200, "Failed to add license file to the model zip file"
+    license_text = r.content.decode("utf-8")
+
+    # find the copyright statements from origin MIT license. It looks like: Copyright (c) {year} {authorname}
+    copyright_statements = re.findall("Copyright.*\n", license_text)[0].strip()
+    huggingface_url = "https://huggingface.co/" + model_id
```
Comment (dhrubo-os) [inline utils/model_uploader/third_party_statements.py:31]: We are giving model source from the UI. Can this model source be dynamic?

File: utils/model_uploader/model_autotracing.py [hunk 2]
```diff
@@ -428,12 +449,15 @@ def main(
     :type pooling_mode: string
     :param model_description: Model description input
     :type model_description: string
+    :param third_party_copyrights_statements: Statements text for non Apache-2.0 licensed third party model. Should be put in the final artifact.
```
Comment (dhrubo-os) [inline utils/model_uploader/model_autotracing.py:452]: are we sure that for any kind non Apahe-2.0 licenses we need this copy right statements? Or this is only for MIT licenses?

File: .github/workflows/model_uploader.yml [hunk 3]
```diff
@@ -96,16 +110,35 @@ jobs:
...
+      run: |
+        if [[ "${{ github.event.inputs.model_license }}" == "MIT" ]]
+        then
+          echo "Uploading MIT licensed model"
+          if [[ "${{ github.event.inputs.MIT_license_url }}" == "" ]]
+          then
+              echo "missing MIT license url"
+              exit 1
+          fi
+          if [[ "${{ github.event.inputs.model_source }}" == "" ]]
```
Comment (dhrubo-os) [inline .github/workflows/model_uploader.yml:124]: what if we provide some other source?

File: opensearch_py_ml/ml_models/sentencetransformermodel.py [hunk 4]
```diff
@@ -657,6 +657,22 @@ def _add_apache_license_to_model_zip_file(self, model_zip_file_path: str):
...
+
+        :param third_party_copyrights_statements: Statements text for non Apache-2.0 licensed third party model. Should be put in the final artifact.
+        :type third_party_copyrights_statements: string
+        :param model_zip_file_path: Path to the model zip file
+        :type model_zip_file_path: string
+        :return: no return value expected
+        :rtype: None
+        """
+        with ZipFile(str(model_zip_file_path), "a") as zipObj:
+            zipObj.writestr("THIRD-PARTY", third_party_copyrights_statements)
```
Comment (dhrubo-os) [inline opensearch_py_ml/ml_models/sentencetransformermodel.py:674]: `THIRD-PARTY` can we put this as a static variable?

File: .github/workflows/model_uploader.yml [hunk 5]
```diff
@@ -49,7 +57,10 @@ on:
         options:
         - "NO"
         - "YES"
-        
+      MIT_license_url:
+        description: "(Optional) MIT license url of the huggingface MIT model."
```
[the remark this answers is not in the document]
Comment (zhichao-aws) [reply .github/workflows/model_uploader.yml:61]: There should be a standard statements for MIT licenses, but the copyright header is customized according to the author info

File: .github/workflows/model_uploader.yml [hunk 6]
```diff
@@ -96,16 +110,35 @@ jobs:
...
+              exit 1
+          fi
+        fi
     - name: Initiate license_line
       id: init_license_line
       run: |
-        echo "verified=:white_check_mark: — It is verified that this model is licensed under Apache 2.0" >> $GITHUB_OUTPUT
-        echo "unverified=- [ ]  :warning: The license cannot be verified. Please confirm by yourself that the model is licensed under Apache 2.0  :warning:" >> $GITHUB_OUTPUT
+        echo "verified_apache=:white_check_mark: — It is verified that this model is licensed under Apache 2.0" >> $GITHUB_OUTPUT
+        echo "verified_mit=:white_check_mark: — It is verified that this model is licensed under MIT, and we have provided copyright statements" >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (zhichao-aws) [reply .github/workflows/model_uploader.yml:134]: I've done a regression test using Apache-2 models and it still works well https://github.com/zhichao-aws/opensearch-py-ml/actions/runs/8843322311/job/24283396611

File: utils/model_uploader/third_party_statements.py (hunk 1 above)
Comment (zhichao-aws) [reply utils/model_uploader/third_party_statements.py:31]: Yes, the model_id and mit_license_url should be provided at the UI chart

File: .github/workflows/model_uploader.yml (hunk 3 above)
Comment (zhichao-aws) [reply .github/workflows/model_uploader.yml:124]: Then the third party statements can be wrong. So we need to check whether they are matched before approve the release

File: opensearch_py_ml/ml_models/sentencetransformermodel.py (hunk 4 above)
Comment (zhichao-aws) [reply opensearch_py_ml/ml_models/sentencetransformermodel.py:674]: Yes, ack

**Comment by zhichao-aws** [comment]
@dhrubo-os Any clue why the IT fails? It seems not related to the change, is it a flaky test?

File: .ci/run-repository.sh [hunk 7]
```diff
@@ -68,11 +68,13 @@ elif [[ "$TASK_TYPE" == "doc" ]]; then
 elif [[ "$TASK_TYPE" == "trace" ]]; then
   # Set up OpenSearch cluster & Run model autotracing (Invoked by model_uploader.yml workflow)
   echo -e "\033[34;1mINFO:\033[0m MODEL_ID: ${MODEL_ID}\033[0m"
+  echo -e "\033[34;1mINFO:\033[0m MODEL_LICENSE: ${MODEL_LICENSE}\033[0m"
   echo -e "\033[34;1mINFO:\033[0m MODEL_VERSION: ${MODEL_VERSION}\033[0m"
   echo -e "\033[34;1mINFO:\033[0m TRACING_FORMAT: ${TRACING_FORMAT}\033[0m"
   echo -e "\033[34;1mINFO:\033[0m EMBEDDING_DIMENSION: ${EMBEDDING_DIMENSION:-N/A}\033[0m"
   echo -e "\033[34;1mINFO:\033[0m POOLING_MODE: ${POOLING_MODE:-N/A}\033[0m"
   echo -e "\033[34;1mINFO:\033[0m MODEL_DESCRIPTION: ${MODEL_DESCRIPTION:-N/A}\033[0m"
+  echo -e "\033[34;1mINFO:\033[0m MIT_LICENSE_URL: ${MIT_LICENSE_URL:-N/A}\033[0m"
```
Comment (dhrubo-os) [inline .ci/run-repository.sh:77]: let's make this "Additional_LICENSE_URL"? In that way may be we can cover other use cases similar to MIT?

File: .github/workflows/model_uploader.yml (hunk 3 above)
Comment (dhrubo-os) [reply .github/workflows/model_uploader.yml:124]: What I tried to mean is let's say `github.event.inputs.model_source` == "XYZ" then this condition will pass, right?

File: opensearch_py_ml/ml_models/sentencetransformermodel.py [hunk 8]
```diff
@@ -38,6 +38,7 @@
 )
 
 LICENSE_URL = "https://github.com/opensearch-project/opensearch-py-ml/raw/main/LICENSE"
+THIRD_PARTY_FILE_NAME = "THIRD-PARTY"
```
Comment (dhrubo-os) [inline opensearch_py_ml/ml_models/sentencetransformermodel.py:41]: Could you please check with the open source engineer that if naming the attribution file as `THIRD-PARTY` looks good to him or not?

File: opensearch_py_ml/ml_models/sentencetransformermodel.py [hunk 9]
```diff
@@ -802,10 +820,16 @@ def save_as_pt(
...
+        :param third_party_copyrights_statements: Statements text for non Apache-2.0 licensed third party model. Should be put in the final artifact.
+        :type third_party_copyrights_statements: string
         :return: model zip file path. The file path where the zip file is being saved
         :rtype: string
         """
 
+        if add_apache_license == True and third_party_copyrights_statements is not None:
+            assert (
+                False
+            ), "When the model is from third party under non Apache-2.0 license, we can not add Apache-2.0 license for it."
```
Comment (dhrubo-os) [inline opensearch_py_ml/ml_models/sentencetransformermodel.py:832]: My understanding is, we will still distribute the artifact under apache 2.0 license but we also need to add extra attribution file to deliberately mention about the contributors name. Am I missing anything?

File: utils/model_uploader/third_party_statements.py (hunk 1 above)
Comment (dhrubo-os) [reply utils/model_uploader/third_party_statements.py:31]: I meant about `huggingface.co`--> Currently we have model source field in the UI. Can we leverage that?

**Comment by dhrubo-os** [comment]
> @dhrubo-os Any clue why the IT fails? It seems not related to the change, is it a flaky test?

Yeah, I think that was a flaky test.

File: opensearch_py_ml/ml_models/sentencetransformermodel.py (hunk 8 above)
Comment (zhichao-aws) [reply opensearch_py_ml/ml_models/sentencetransformermodel.py:41]: The name `THIRD_PARTY` is provided by the open source engineer

File: .ci/run-repository.sh (hunk 7 above)
Comment (zhichao-aws) [reply .ci/run-repository.sh:77]: Makes sense. Ack

File: .github/workflows/model_uploader.yml [hunk 10]
```diff
@@ -49,7 +57,10 @@ on:
         options:
         - "NO"
         - "YES"
-        
+      MIT_license_url:
```
[the remark this answers is not in the document]
Comment (zhichao-aws) [reply .github/workflows/model_uploader.yml:60]: Ack

**Comment by zhichao-aws** [comment]
close it for the issue of BGE training data. Feel free to re-open it when we need to upload other MIT models

**Pull request closed without merging**